server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentTracker.java

```diff
@@ -0,0 +1,425 @@
+        localRefreshTimeMs = currentTimeMs;
+        remoteRefreshTimeMs = currentTimeMs;
+        uploadBytesMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesMovingAverageWindowSize));
+        uploadBytesPerSecMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesPerSecMovingAverageWindowSize));
+        uploadTimeMsMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadTimeMsMovingAverageWindowSize));
+    }
+
+    ShardId getShardId() {
+        return shardId;
+    }
+
+    long getLocalRefreshSeqNo() {
+        return localRefreshSeqNo;
+    }
+
+    void updateLocalRefreshSeqNo(long localRefreshSeqNo) {
+        assert localRefreshSeqNo > this.localRefreshSeqNo : "newLocalRefreshSeqNo="
+            + localRefreshSeqNo
+            + ">="
+            + "currentLocalRefreshSeqNo="
+            + this.localRefreshSeqNo;
+        this.localRefreshSeqNo = localRefreshSeqNo;
+        computeSeqNoLag();
+    }
+
+    long getLocalRefreshTimeMs() {
+        return localRefreshTimeMs;
+    }
+
+    void updateLocalRefreshTimeMs(long localRefreshTimeMs) {
+        assert localRefreshTimeMs > this.localRefreshTimeMs : "newLocalRefreshTimeMs="
+            + localRefreshTimeMs
+            + ">="
+            + "currentLocalRefreshTimeMs="
+            + this.localRefreshTimeMs;
+        this.localRefreshTimeMs = localRefreshTimeMs;
+        computeTimeMsLag();
+    }
+
+    long getRemoteRefreshSeqNo() {
+        return remoteRefreshSeqNo;
+    }
+
+    void updateRemoteRefreshSeqNo(long remoteRefreshSeqNo) {
+        assert remoteRefreshSeqNo > this.remoteRefreshSeqNo : "newRemoteRefreshSeqNo="
+            + remoteRefreshSeqNo
+            + ">="
+            + "currentRemoteRefreshSeqNo="
+            + this.remoteRefreshSeqNo;
+        this.remoteRefreshSeqNo = remoteRefreshSeqNo;
+        computeSeqNoLag();
+    }
+
+    long getRemoteRefreshTimeMs() {
+        return remoteRefreshTimeMs;
+    }
+
+    void updateRemoteRefreshTimeMs(long remoteRefreshTimeMs) {
+        assert remoteRefreshTimeMs > this.remoteRefreshTimeMs : "newRemoteRefreshTimeMs="
+            + remoteRefreshTimeMs
+            + ">="
+            + "currentRemoteRefreshTimeMs="
+            + this.remoteRefreshTimeMs;
+        this.remoteRefreshTimeMs = remoteRefreshTimeMs;
+        computeTimeMsLag();
+    }
+
+    private void computeSeqNoLag() {
+        seqNoLag = localRefreshSeqNo - remoteRefreshSeqNo;
+    }
+
+    long getSeqNoLag() {
+        return seqNoLag;
+    }
+
+    private void computeTimeMsLag() {
+        timeMsLag = localRefreshTimeMs - remoteRefreshTimeMs;
+    }
+
+    long getTimeMsLag() {
+        return timeMsLag;
+    }
+
+    long getBytesLag() {
+        return bytesLag;
+    }
+
+    long getUploadBytesStarted() {
+        return uploadBytesStarted;
+    }
+
+    void addUploadBytesStarted(long size) {
+        uploadBytesStarted += size;
```

Comment:
Sorry, I actually missed all these increment operations. I think this is a valid reason to use an AtomicLong because this isn't an atomic operation for a volatile field and isn't truly thread safe unless you use an AtomicLong (or protect it with a lock). The fields that are only read or written with a single atomic operation can be volatile primitives, but these increment and add operations need AtomicLong or external synchronization.

Reply:
Only one thread at a time can increment the fields with volatile accessibility. The fields that can be modified concurrently are still AtomicLong / Concurrent extensions.

Reply:
So you're saying something external to this class ensures single threaded access to these methods?

Reply:
Yes, that's true for most of the fields.